src/panels/lovelace/common/directives/long-press-directive.ts

```diff
@@ -0,0 +1,131 @@
+import { directive, PropertyPart } from 'lit-html';
+import "@material/mwc-ripple";
+
+const isTouch =
+  "ontouchstart" in window ||
+  navigator.maxTouchPoints > 0 ||
+  navigator.msMaxTouchPoints > 0;
+
+interface LongPress extends HTMLElement {
+  holdTime: number;
+  bind(element:Element): void;
+}
+interface LongPressElement extends Element {
+  longPress?: boolean;
+}
+
+class LongPress extends HTMLElement implements LongPress{
+  public holdTime: number;
+  protected ripple: any;
+  protected timer: number|undefined;
+  protected held: boolean;
+
+  constructor() {
+    super();
+    this.holdTime = 500;
+    this.ripple = document.createElement("mwc-ripple");
+    this.timer = undefined;
+    this.held = false;
+  };
+
+  public connectedCallback() {
+    Object.assign(this.style, {
+      position: "absolute",
+      width: isTouch ? "100px" : "50px",
+      height: isTouch ? "100px" : "50px",
+      transform: "translate(-50%, -50%)",
+      pointerEvents: "none",
+    });
+
+    this.appendChild(this.ripple);
+    this.ripple.primary = true;
+
+    [
+      isTouch ? "touchcancel" : "mouseout",
+      "mouseup",
+      "touchmove",
+      "mousewheel",
+      "wheel",
+      "scroll",
+    ].forEach((ev) => {
+      document.addEventListener(ev, () => {
+        clearTimeout(this.timer);
+        this.stopAnimation();
+      }, {passive: true});
+    });
+  }
+
+  public bind(element: LongPressElement) {
+    if(element.longPress) { return; }
+    element.longPress = true;
+    element.addEventListener(isTouch ? "touchstart" : "mousedown",
+      (ev: Event) => {
+        this.held = false;
+        let x;
+        let y;
+        if((ev as TouchEvent).touches)
+        {
+          x = (ev as TouchEvent).touches[0].pageX;
+          y = (ev as TouchEvent).touches[0].pageY;
+        } else {
+          x = (ev as MouseEvent).pageX;
+          y = (ev as MouseEvent).pageY;
+        }
+        this.timer = window.setTimeout(() => {
+          this.startAnimation(x, y);
+          this.held = true;
+        }, this.holdTime);
+      }, {passive: true});
+    element.addEventListener("click",
+      () => {
+        this.stopAnimation();
+        if(this.held) {
+          element.dispatchEvent(new Event('ha-hold'));
+        } else {
+          element.dispatchEvent(new Event('ha-click'));
+        }
+      });
+  }
+
+  private startAnimation(x:number, y:number) {
+    Object.assign(this.style, {
+      left: `${x}px`,
+      top: `${y}px`,
+      visibility: "visible",
+    });
+    this.ripple.disabled = false;
+    this.ripple.active = true;
+    this.ripple.unbounded = true;
+  }
+
+  private stopAnimation() {
+    this.ripple.active = false;
+    this.ripple.disabled = true;
+    this.style.visibility = "hidden";
```

Comment:
should this be visibility or display: none.

Reply:
display: none feels more right. I'll change it.